### PR TITLE
[K8s-Gen] Kube Map

### PR DESCRIPTION
**Pull request by Lewiky**
Adds the mapping between `Templefile` and `KubeGen`. - Most of the diff is adding all of the yaml test files for the `ProjectBuilder` Unit and E2E tests, can probably be skipped

### CODE REVIEW

File: src/e2e/scala/temple/SimpleE2ETest.scala [hunk 1]
```diff
@@ -24,8 +23,8 @@ class SimpleE2ETest extends FlatSpec with Matchers {
       ),
     )
 
-    // Two folders should have been generated
-    val expectedFolders = Set("templeuser-db", "templeuser").map(dir => basePath.resolve(dir))
+    // Four folders should have been generated
```
Comment (c-harding) [inline src/e2e/scala/temple/SimpleE2ETest.scala]:
```suggestion
    // Exactly these folders should have been generated
```

Prevents them getting out of sync, and clarifies that it’s not just the number of folders, we are checking which ones are generated

File: src/main/scala/temple/builder/KubernetesBuilder.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package temple.builder
+
+import temple.DSL.semantics.Metadata.Database
+import temple.DSL.semantics.{Metadata, ServiceBlock}
```
Comment (c-harding) [inline src/main/scala/temple/builder/KubernetesBuilder.scala]:
```suggestion
import temple.ast.Metadata.Database
import temple.ast.{Metadata, ServiceBlock}
```
The merge 6e84646 broke this

File: src/e2e/scala/temple/SimpleE2ETest.scala [hunk 3]
```diff
@@ -24,8 +23,8 @@ class SimpleE2ETest extends FlatSpec with Matchers {
...
+    val expectedFolders = Set("templeuser-db", "templeuser", "kong", "kube").map(dir => basePath.resolve(dir))
```
Comment (jaylees14) [inline src/e2e/scala/temple/SimpleE2ETest.scala:28]: Should we not be checking the content of these folders?

File: src/main/scala/temple/builder/project/ProjectBuilder.scala [hunk 4]
```diff
@@ -35,6 +37,16 @@ object ProjectBuilder {
         (File(s"${name.toLowerCase}", "Dockerfile"), dockerfileContents)
     }
 
-    Project(databaseCreationScripts ++ dockerfiles)
+    val kubeFiles = KubernetesGenerator.generate(
+      KubernetesBuilder
+        .createServiceKubeFiles(
+          templefile.projectName,
+          (templefile.services.zip(1024 until Int.MaxValue) map {
```
Comment (c-harding) [inline src/main/scala/temple/builder/project/ProjectBuilder.scala]: Adding a dot before map means you don’t need the surrounding parens, may make it nicer

File: src/main/scala/temple/builder/project/ProjectBuilder.scala [hunk 5]
```diff
@@ -35,6 +37,16 @@ object ProjectBuilder {
         (File(s"${name.toLowerCase}", "Dockerfile"), dockerfileContents)
     }
 
-    Project(databaseCreationScripts ++ dockerfiles)
+    val kubeFiles = KubernetesGenerator.generate(
+      KubernetesBuilder
```
Comment (jaylees14) [inline src/main/scala/temple/builder/project/ProjectBuilder.scala]: Can we follow the pattern above and separate these into two variables?

File: src/main/scala/temple/builder/project/ProjectBuilder.scala (hunk 4 above)
Comment (jaylees14) [inline src/main/scala/temple/builder/project/ProjectBuilder.scala]: Also probably worth using the same one as above?

File: src/main/scala/temple/generate/kube/KubernetesGenerator.scala [hunk 6]
```diff
@@ -91,7 +91,7 @@ object KubernetesGenerator {
       service.dbImage,
       name,
       ports = Seq(),
-      env = service.envVars.map(x => EnvVar(x._1, x._2)),
+      env = service.dbEnvVars.map(x => EnvVar(x._1, x._2)),
```
Comment (c-harding) [inline src/main/scala/temple/generate/kube/KubernetesGenerator.scala]: Why doesn’t Service already contain EnvVars rather than tuples?

File: src/main/scala/temple/generate/kube/ast/OrchestrationType.scala [hunk 7]
```diff
@@ -7,16 +7,18 @@ object OrchestrationType {
...
+    * @param dbImage            The name of the docker image to use for the service's datastore
+    * @param ports              The ports the Pod should expose, including the names of the ports (i.e www -> 80)
+    * @param replicas           The number of replicas of the pod that should be exposed
+    * @param secretName         The name of the Kubernetes secret used to fetch images from the registry
+    * @param appEnvVars         A sequence of key -> value to set as environment variables in the app container
+    * @param dbEnvVars          A sequence of key -> value to set as environment variables in the db container
+    * @param dbStorage          Object that defines how the service db should store data
     * @param dbLifecycleCommand The command to be ran in the database container on startup e.g to init the db
-    * @param usesAuth Whether this service requires authentication with the API Gateway before requests
+    * @param usesAuth           Whether this service requires authentication with the API Gateway before requests
```
Comment (c-harding) [inline src/main/scala/temple/generate/kube/ast/OrchestrationType.scala:21]: Is this spacing normal?

File: src/test/scala/temple/builder/project/ProjectBuilderTest.scala [hunk 8]
```diff
@@ -9,33 +9,36 @@ class ProjectBuilderTest extends FlatSpec with Matchers {
...
     val project = ProjectBuilder.build(ProjectBuilderTestData.simpleTemplefile)
-    project.files shouldBe Map(
-      File("templeuser-db", "init.sql") -> ProjectBuilderTestData.simpleTemplefilePostgresCreateOutput,
-      File("templeuser", "Dockerfile")  -> ProjectBuilderTestData.simpleTemplefileUsersDockerfile,
-    )
+    val expected = Map(
+        File("templeuser-db", "init.sql") -> ProjectBuilderTestData.simpleTemplefilePostgresCreateOutput,
+        File("templeuser", "Dockerfile")  -> ProjectBuilderTestData.simpleTemplefileUsersDockerfile,
+      ) ++ ProjectBuilderTestData.simpleTemplefileKubeScripts
+    project.files.toSet shouldBe expected.toSet
```
Comment (c-harding) [inline src/test/scala/temple/builder/project/ProjectBuilderTest.scala]: Does this work without the toSet?

**Review by c-harding — APPROVED** [review]
Few comments, looks nice

File: src/test/scala/temple/builder/project/ProjectBuilderTest.scala [hunk 9]
```diff
@@ -9,33 +9,36 @@ class ProjectBuilderTest extends FlatSpec with Matchers {
...
   it should "correctly create a simple project using postgres as the default" in {
     val project = ProjectBuilder.build(ProjectBuilderTestData.simpleTemplefile)
-    project.files shouldBe Map(
-      File("templeuser-db", "init.sql") -> ProjectBuilderTestData.simpleTemplefilePostgresCreateOutput,
-      File("templeuser", "Dockerfile")  -> ProjectBuilderTestData.simpleTemplefileUsersDockerfile,
-    )
+    val expected = Map(
+        File("templeuser-db", "init.sql") -> ProjectBuilderTestData.simpleTemplefilePostgresCreateOutput,
+        File("templeuser", "Dockerfile")  -> ProjectBuilderTestData.simpleTemplefileUsersDockerfile,
+      ) ++ ProjectBuilderTestData.simpleTemplefileKubeScripts
```
Comment (jaylees14) [inline src/test/scala/temple/builder/project/ProjectBuilderTest.scala]: Hmm not sure I'm a fan of this, would prefer to explicitly list every file - with its contents being the thing in `TestData`

File: src/test/scala/temple/builder/project/ProjectBuilderTestData.scala [hunk 10]
```diff
@@ -93,6 +96,184 @@ object ProjectBuilderTestData {
...
+        |  labels:
+        |    app: TempleUser
+        |    kind: service
+        |spec:
+        |  replicas: 1
+        |  selector:
+        |    matchLabels:
+        |      app: TempleUser
+        |      kind: service
+        |  template:
```
Comment (jaylees14) [inline src/test/scala/temple/builder/project/ProjectBuilderTestData.scala]: Yeah this is quite hard to follow - can we break it into separate variables for each file's content

File: src/main/scala/temple/generate/kube/KubernetesGenerator.scala (hunk 6 above)
Comment (Lewiky) [reply src/main/scala/temple/generate/kube/KubernetesGenerator.scala]: The way I saw it - the `kube.Spec.EnvVar` here have a certain JSON encoding defined on them that's pretty specific to KubeGen, whereas the `OrchestrationType.Service` above should be generic to all orchestration

File: src/main/scala/temple/generate/kube/ast/OrchestrationType.scala (hunk 7 above)
Comment (Lewiky) [reply src/main/scala/temple/generate/kube/ast/OrchestrationType.scala:21]: Think that's how scalafmt wants them

File: src/main/scala/temple/builder/project/ProjectBuilder.scala [hunk 11]
```diff
@@ -35,6 +37,16 @@ object ProjectBuilder {
...
+        .createServiceKubeFiles(
```
Comment (c-harding) [inline src/main/scala/temple/builder/project/ProjectBuilder.scala]: Please could this go on the same line as `KubernetesBuilder`, the chaining doesn’t make sense if there isn’t a flow of information (KubernetesBuilder isn’t some data to be passed down the pipeline)

File: src/main/scala/temple/builder/project/ProjectBuilder.scala [hunk 12]
```diff
@@ -41,9 +40,12 @@ object ProjectBuilder {
...
         .createServiceKubeFiles(
           templefile.projectName,
-          (templefile.services.zip(1024 until Int.MaxValue) map {
-            case ((name, service), port) => (name, service, port)
-          }).toSeq,
+          templefile.services
+            .zip(1024 until Int.MaxValue)
+            .map {
+              case ((name, service), port) => (name, service, port)
+            }
```
Comment (c-harding) [inline src/main/scala/temple/builder/project/ProjectBuilder.scala]: Could be one-lined

Comment (xsanda) [reply src/main/scala/temple/builder/project/ProjectBuilder.scala]: In fact, having this defined on the Templefile might be useful
```scala
def servicesWithPorts: Iterable[(String, ServiceBlock, Int)] =
  templefile.services
    .zip(1024 until Int.MaxValue)
    .map { case ((name, service), port) => (name, service, port) }
```

File: src/main/scala/temple/builder/project/ProjectBuilder.scala [hunk 13]
```diff
@@ -35,6 +36,19 @@ object ProjectBuilder {
...
-    Project(databaseCreationScripts ++ dockerfiles)
+    val kubeFiles = KubernetesGenerator.generate(
+      KubernetesBuilder
+        .createServiceKubeFiles(
+          templefile.projectName,
+          templefile.services
+            .zip(1024 until Int.MaxValue)
+            .map {
+              case ((name, service), port) => (name, service, port)
+            }
```
Comment (c-harding) [inline src/main/scala/temple/builder/project/ProjectBuilder.scala]: In fact, having this defined on the Templefile might be useful, to avoid reuse of the magic number 1024
```scala
def servicesWithPorts: Iterable[(String, ServiceBlock, Int)] =
  templefile.services
    .zip(1024 until Int.MaxValue)
    .map { case ((name, service), port) => (name, service, port) }
```

File: src/main/scala/temple/generate/kube/ast/OrchestrationType.scala (hunk 7 above)
Comment (c-harding) [reply src/main/scala/temple/generate/kube/ast/OrchestrationType.scala:21]: Mine doesn’t, but sure

File: src/main/scala/temple/generate/kube/KubernetesGenerator.scala (hunk 6 above)
Comment (c-harding) [reply src/main/scala/temple/generate/kube/KubernetesGenerator.scala]: Ok, fair enough. Worth noting that `service.dbEnvVars.map(EnvVar.tupled)` is equivalent, and is arguably nicer than listing each component with `_i`

File: src/main/scala/temple/builder/project/ProjectBuilder.scala (hunk 13 above)
Comment (c-harding) [reply src/main/scala/temple/builder/project/ProjectBuilder.scala]: Also wondering about `Iterator from 1024`/`Iterator.from(1024)` instead of explicitly stating the max value

File: src/main/scala/temple/builder/project/ProjectBuilder.scala [hunk 14]
```diff
@@ -28,13 +29,23 @@ object ProjectBuilder {
         }
     }
 
-    val dockerfiles = templefile.services.zip(1024 until Int.MaxValue).map {
+    val servicePortIterator = templefile.services.zip(1024 until Int.MaxValue)
```
Comment (c-harding) [inline src/main/scala/temple/builder/project/ProjectBuilder.scala]: This feels like something that should be done in the main code, and not in tests. See the suggested code snippet in the discussion tab (it was attached to outdated code)

**Comment by Lewiky** [comment]
Boy did this get out of hand

**Review by c-harding — APPROVED** [review]

File: src/e2e/scala/temple/SimpleE2ETest.scala [hunk 15]
```diff
@@ -43,5 +43,65 @@ class SimpleE2ETest extends FlatSpec with Matchers {
...
+    val templeKongKubeService = Files.readString(basePath.resolve("kube/kong").resolve("kong-service.yaml"))
+    templeKongKubeService shouldBe FileUtils.readResources("kube/kong/kong-service.yaml")
+
+    val templeKongKubeDbService = Files.readString(basePath.resolve("kube/kong").resolve("kong-db-service.yaml"))
+    templeKongKubeDbService shouldBe FileUtils.readResources("kube/kong/kong-db-service.yaml")
+
+    val templeKongKubeMigration = Files.readString(basePath.resolve("kube/kong").resolve("kong-migration-job.yaml"))
+    templeKongKubeMigration shouldBe FileUtils.readResources("kube/kong/kong-migration-job.yaml")
+
+    // Only these files should be present in the kube/TempleUser folder
```
Comment (jaylees14) [inline src/e2e/scala/temple/SimpleE2ETest.scala]: We may want to lower case the `TempleUser`, to be consistent with the service folder name?

File: src/main/scala/temple/builder/OrchestrationBuilder.scala [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+object OrchestrationBuilder {
+
+  def createServiceOrchestrationRoot(
+    projectName: String,
+    services: Seq[(String, ServiceBlock, Int)],
+  ): OrchestrationRoot =
+    OrchestrationRoot(
+      services map {
+        case (name, service, port) =>
+          val dockerImage = s"temple-$projectName-$name"
```
Comment (jaylees14) [inline src/main/scala/temple/builder/OrchestrationBuilder.scala]: Not sure I like adding `temple` to the start of each Docker image - feels like it ties it too closely with generated code? @xsanda @smithwjv?

File: src/main/scala/temple/builder/project/ProjectBuilder.scala [hunk 17]
```diff
@@ -28,13 +29,19 @@ object ProjectBuilder {
...
-      case ((name, service), port) =>
+    val dockerfiles = templefile.servicesWithPorts.map {
+      case (name, service, port) =>
         val dockerfileRoot     = DockerfileBuilder.createServiceDockerfile(name.toLowerCase, service, port)
         val dockerfileContents = DockerfileGenerator.generate(dockerfileRoot)
         (File(s"${name.toLowerCase}", "Dockerfile"), dockerfileContents)
     }
 
-    Project(databaseCreationScripts ++ dockerfiles)
+    val orchestrationRoot = OrchestrationBuilder.createServiceOrchestrationRoot(
```
Comment (jaylees14) [inline src/main/scala/temple/builder/project/ProjectBuilder.scala:40]: Just a consistency side note - in the `database` and `dockerfile` cases we map over the services in this file, whereas with Kubernetes we map over in the builder - any particular reason why?

File: src/main/scala/temple/generate/kube/KubernetesGenerator.scala (hunk 6 above)
Comment (jaylees14) [inline src/main/scala/temple/generate/kube/KubernetesGenerator.scala]: Can we not also use `.map(EnvVar.tupled)`?

File: src/e2e/scala/temple/SimpleE2ETest.scala (hunk 15 above)
Comment (jaylees14) [reply src/e2e/scala/temple/SimpleE2ETest.scala]: I realise this is quite a few changes 😆, but we should probably be consistent

**Review by jaylees14 — COMMENTED** [review]
Just a few more comments, but overall looks good. LMK if you don't agree

File: src/main/scala/temple/builder/OrchestrationBuilder.scala (hunk 16 above)
Comment (c-harding) [reply src/main/scala/temple/builder/OrchestrationBuilder.scala]: Is there a reason we didn’t just go for `s"$projectName-$name"`? @Lewiky @jaylees14

File: src/main/scala/temple/builder/project/ProjectBuilder.scala (hunk 17 above)
Comment (Lewiky) [reply src/main/scala/temple/builder/project/ProjectBuilder.scala:40]: The kube gen acts on all services at once (This way shared file i.e kong get generated properly)

**Review by jaylees14 — APPROVED** [review]
I'll just reiterate what we said in person: I don't really agree with Kubernetes being a single generator - I think it should be a per service thing, with the kong configuration added on top. This way it feels more modular, in the sense that if we wanted to switch API gateways in the future, it would be a single generator change, rather than having to dig through the entirety of the Kube generator.
